presto-main/src/main/java/com/facebook/presto/operator/PageSupplier.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.annotation.Nullable;
+
+public interface PageSupplier
```

Comment:
I think some kind of `Iterators#mergeSorted` that supports `BlockingIterator<T>` would need to be implemented.

Alternatively, we could have a special class that stores either value or blocked future:
```
class ValueOrBlocked<T> {
  ...
  Optional<T> getValue();
  Future<?> isBlocked();
}
```
and use it in iterators:
```
// page supplier
Iterator<ValueOrBlocked<Page>> ...
// page with position supplier
Iterator<ValueOrBlocked<PageWithPosition>> ...

// use it in merging with Guava (blocked ValueOrBlocked have precedence over non-blocked values)
Iterator<ValueOrBlocked<PageWithPosition>> Iterators.mergeSorted(pageWithPositionIterators);
```
think this approach is more hacky, but allows us to reuse Guava `mergeSorted`